lib/Runtime/Base/FunctionBody.cpp

```diff
@@ -8305,30 +8305,45 @@ namespace Js
 
     PropertyGuard* EntryPointInfo::RegisterSharedPropertyGuard(Js::PropertyId propertyId, ScriptContext* scriptContext)
     {
-        AutoCriticalSection autoCs(FunctionProxy::GetLock());
-        if (this->sharedPropertyGuards == nullptr)
+        auto localGuards = this->sharedPropertyGuards;
+        if (localGuards == nullptr)
         {
             Recycler* recycler = scriptContext->GetRecycler();
-            this->sharedPropertyGuards = RecyclerNew(recycler, SharedPropertyGuardDictionary, recycler);
+            localGuards = RecyclerNew(recycler, SharedPropertyGuardDictionary, recycler);
         }
 
         PropertyGuard* guard;
-        if (!this->sharedPropertyGuards->TryGetValue(propertyId, &guard))
+        if (!localGuards->TryGetValue(propertyId, &guard))
         {
             ThreadContext* threadContext = scriptContext->GetThreadContext();
             guard = threadContext->RegisterSharedPropertyGuard(propertyId);
-            this->sharedPropertyGuards->Add(propertyId, guard);
+
+            struct AutoResizeLock
+            {
+                AutoResizeLock(SharedPropertyGuardDictionary* d) :d(d) { d->LockResize(); }
```

Comment:
you are correct. this is recycler allocated, should not worry the old buffer being deleted